File: packages/mirascript/src/vm/lib/global/sequence/flatten.ts
import { VmLib, expectArray, expectNumber } from '../../helpers.js';

export const flatten = VmLib(
    (data, depth = 1) => {
        expectArray('data', data, data);
        return data.flat(expectNumber('depth', depth) as 1);
    },
    {
        summary: '将数组扁平化',
        params: {
            data: { type: 'array', description: '要扁平化的数组' },
            depth: { type: 'number | nil', description: '扁平化的深度，默认为 1' },
        },
        returns: { type: 'array' },
        examples: ['flatten([[1, 2], [3, [4]]], 2) // [1, 2, 3, 4]'],
    },
);

File: crates/python/mirascript/_vm/lib/vm_global/sequence/flatten.py
from ....types import Uninitialized, VmAny
from ..._helpers import _expect_array, _expect_number


def flatten(data: VmAny = Uninitialized, depth: VmAny = 1):
    _expect_array("data", data, data)

    def flat(arr, d):
        if d < 1:
            return arr
        result = []
        for item in arr:
            if isinstance(item, list):
                result.extend(flat(item, d - 1))
            else:
                result.append(item)
        return result

    d = _expect_number("depth", depth)
    return flat(data, d)
